polkadot/xcm/xcm-fee-payment-runtime-api/tests/mock.rs

```diff
@@ -0,0 +1,497 @@
+
+	let mut ext = sp_io::TestExternalities::new(t);
+	ext.execute_with(|| System::set_block_number(1));
+	ext
+}
+
+pub fn new_test_ext_with_balances_and_assets(
+	balances: Vec<(AccountId, Balance)>,
+	assets: Vec<(AssetIdForAssetsPallet, AccountId, Balance)>,
+) -> sp_io::TestExternalities {
+	let mut t = frame_system::GenesisConfig::<TestRuntime>::default().build_storage().unwrap();
+
+	pallet_balances::GenesisConfig::<TestRuntime> { balances }
+		.assimilate_storage(&mut t)
+		.unwrap();
+
+	pallet_assets::GenesisConfig::<TestRuntime> {
+		assets: vec![
+			// id, owner, is_sufficient, min_balance.
+			// We don't actually need this to be sufficient, since we use the native assets in
+			// tests for the existential deposit.
+			(1, 0, true, 1),
+		],
+		metadata: vec![
+			// id, name, symbol, decimals.
+			(1, "Relay Token".into(), "RLY".into(), 12),
+		],
+		accounts: assets,
+	}
+	.assimilate_storage(&mut t)
+	.unwrap();
+
+	let mut ext = sp_io::TestExternalities::new(t);
+	ext.execute_with(|| System::set_block_number(1));
+	ext
+}
+
+#[derive(Clone)]
+pub(crate) struct TestClient;
+
+pub(crate) struct RuntimeApi {
+	_inner: TestClient,
+}
+
+impl sp_api::ProvideRuntimeApi<Block> for TestClient {
+	type Api = RuntimeApi;
+	fn runtime_api(&self) -> sp_api::ApiRef<Self::Api> {
+		RuntimeApi { _inner: self.clone() }.into()
+	}
+}
+
+sp_api::mock_impl_runtime_apis! {
+	impl XcmPaymentApi<Block> for RuntimeApi {
+		fn query_acceptable_payment_assets(xcm_version: XcmVersion) -> Result<Vec<VersionedAssetId>, XcmPaymentApiError> {
+			if xcm_version != 4 { return Err(XcmPaymentApiError::UnhandledXcmVersion) };
+			Ok(vec![VersionedAssetId::V4(HereLocation::get().into())])
+		}
+
+		fn query_xcm_weight(message: VersionedXcm<()>) -> Result<Weight, XcmPaymentApiError> {
+			XcmPallet::query_xcm_weight(message)
+		}
+
+		fn query_weight_to_asset_fee(weight: Weight, asset: VersionedAssetId) -> Result<u128, XcmPaymentApiError> {
+			let local_asset = VersionedAssetId::V4(HereLocation::get().into());
+			let asset = asset
+				.into_version(4)
+				.map_err(|_| XcmPaymentApiError::VersionedConversionFailed)?;
+
+			if asset != local_asset { return Err(XcmPaymentApiError::AssetNotFound); }
+
+			Ok(WeightToFee::weight_to_fee(&weight))
+		}
+
+		fn query_delivery_fees(destination: VersionedLocation, message: VersionedXcm<()>) -> Result<VersionedAssets, XcmPaymentApiError> {
+			XcmPallet::query_delivery_fees(destination, message)
+		}
+	}
+
+	impl XcmDryRunApi<Block, RuntimeCall> for RuntimeApi {
+		fn dry_run_extrinsic(extrinsic: <Block as BlockT>::Extrinsic) -> Result<XcmDryRunEffects, ()> {
+			// We want to record the XCM that's executed, so we can return it.
+			pallet_xcm::ShouldRecordXcm::<TestRuntime>::put(true);
+			// Asserting only because it's a test.
+			assert_ok!(Executive::apply_extrinsic(extrinsic));
+			// Turning recording off after we execute the extrinsic just in case.
+			pallet_xcm::ShouldRecordXcm::<TestRuntime>::put(false);
```

Comment:
hmm, not sure if we should "set this example"

maybe we can replace this with some documentation describing how one advantage of using runtime apis for dry-running is nothing gets committed to storage anyway